### PR TITLE
Fixed #4772 create unique connection for each notebook by using noteb…

### LINKED ISSUE #4772: Multiple SQL Notebooks with same query running against same database and table hang
Issue Type: <b>Bug</b>

image

REPO
1. Create a SQL Notebook & write a query against a table in the DB
SELECT TOP (1000) [InvoiceLineId]
      ,[InvoiceId]
      ,[TrackId]
      ,[UnitPrice]
      ,[Quantity]
  FROM [dbo].[InvoiceLine]

2. Create a second notebook and run the same query
SELECT TOP (1000) [InvoiceLineId]
      ,[InvoiceId]
      ,[TrackId]
      ,[UnitPrice]
      ,[Quantity]
  FROM [dbo].[InvoiceLine]

3. Note that the queries hang

Azure Data Studio version: Azure Data Studio 1.5.2 (f74080c96310354bc92211f1826c0fdce78ca2f6, 2019-03-22T06:09:10.602Z)
OS version: Darwin x64 18.2.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) M-5Y51 CPU @ 1.10GHz (4 x 1200)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: enabled<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|8.00GB (0.10GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

</details>Extensions: none

### PR DESCRIPTION
**Pull request by YurongHe**
Fixed #4772 create unique connection for each notebook by using notebook path as uri
Disconnect sqlConnection from SqlKernel to ensure no connection left after the notebook is closed.

### CODE REVIEW

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 1]
```diff
@@ -46,8 +45,14 @@ export interface SQLData {
 }
 
 export class SqlSessionManager implements nb.SessionManager {
+	private _session: nb.ISession;
+
 	constructor(private _instantiationService: IInstantiationService) { }
 
+	public get id(): string {
+		return this._session.id;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: This should check that _session isn't undefined

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 2]
```diff
@@ -65,11 +70,15 @@ export class SqlSessionManager implements nb.SessionManager {
...
 
 	startNew(options: nb.ISessionOptions): Thenable<nb.ISession> {
-		let session = new SqlSession(options, this._instantiationService);
-		return Promise.resolve(session);
+		this._session = new SqlSession(options, this._instantiationService);
+		return Promise.resolve(this._session);
 	}
 
 	shutdown(id: string): Thenable<void> {
+		if (this._session.kernel) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: _session can be undefined

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 3]
```diff
@@ -65,11 +70,15 @@ export class SqlSessionManager implements nb.SessionManager {
 	}
 
 	startNew(options: nb.ISessionOptions): Thenable<nb.ISession> {
-		let session = new SqlSession(options, this._instantiationService);
-		return Promise.resolve(session);
+		this._session = new SqlSession(options, this._instantiationService);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: What about cleaning up if a _session is already defined?

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 4]
```diff
@@ -65,11 +70,15 @@ export class SqlSessionManager implements nb.SessionManager {
...
+		return Promise.resolve(this._session);
 	}
 
 	shutdown(id: string): Thenable<void> {
+		if (this._session.kernel) {
+			let sqlKernel = this._session.kernel as SqlKernel;
+			return sqlKernel.disconnect();
+		}
 		return Promise.resolve();
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: Clear _session here?

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 5]
```diff
@@ -311,6 +325,13 @@ class SqlKernel extends Disposable implements nb.IKernel {
 		}
 		// TODO issue #2746 should ideally show a warning inside the dialog if have no data
 	}
+
+	public async disconnect(): Promise<void> {
+		if (this._path) {
+			await this._connectionManagementService.disconnect(this._path);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: What's path here? Can you have multiple kernels with the same path? If so disconnecting that connection might break other notebooks right?

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 6]
```diff
@@ -210,6 +221,10 @@ class SqlKernel extends Disposable implements nb.IKernel {
 		return info;
 	}
 
+	public set path(value: string) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: nit: Could also just have this be a parameter on the constructor

**Comment by YurongHe** [comment]
> What happens when the path changes (i.e. we go from untitled to saved notebook)? Does the connection still work?

It will be a new doc from untitled to saved. So it created a new connection when run

**Comment by YurongHe** [comment]
Not sure why it was closed. Should be open

File: src/sql/workbench/services/notebook/sql/sqlSessionManager.ts [hunk 7]
```diff
@@ -146,11 +161,12 @@ class SqlKernel extends Disposable implements nb.IKernel {
...
 	private _magicToExecutorMap = new Map<string, ExternalScriptMagic>();
 
-	constructor(@IConnectionManagementService private _connectionManagementService: IConnectionManagementService,
+	constructor(private _path: string,
+		@IConnectionManagementService private _connectionManagementService: IConnectionManagementService,
 		@ICapabilitiesService private _capabilitiesService: ICapabilitiesService,
 		@IInstantiationService private _instantiationService: IInstantiationService,
 		@IErrorMessageService private _errorMessageService: IErrorMessageService,
-		@IConfigurationService private _configurationService: IConfigurationService
+		@IConfigurationService private _configurationService: IConfigurationService,
```
Comment (chlafreniere) [inline src/sql/workbench/services/notebook/sql/sqlSessionManager.ts]: nit: extra comma

**Pull request merged**